Cram3.1 - RANS Nx16 Order 0

yash-puligundla opened the pull request:
### Description

Work in Progress!!
The purpose of this PR is to facilitate easier Code Review. 

@cmnbroad I have added an abstract class for RANS called "RANSNew". The existing RANS class has not been deleted yet as it is being used at other places.

I am still in the middle of refactoring "RANSTest" class, specifically, testBuffersMeetBoundaryExpectations and testRansHeader methods. But, the rest is ready for your review.

cmnbroad commented on src/test/java/htsjdk/samtools/cram/CRAMCodecCorpusTest.java:
```diff
@@ -0,0 +1,244 @@
+    // RANS tests
+    /////////////////////////////////////////////////////////////////////////////////////////////////
+
+    //TODO: the TestDataProviders tests fail if the hts codecs corpus isn't available because
```
Remove the `@Test` annotation from testGetHTSCodecsCorpus, and then I think the "dependsOn" methods will kick in and cause the downstream tests to be skipped if no corpus is available.

yash-puligundla replied:
It's actually the other way. @Test annotation is required to skip the dependent tests.

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16Params.java:
```diff
@@ -0,0 +1,67 @@
+        return ORDER.fromInt(format & ORDER_FLAG); //convert into order type
+    }
+
+    protected int getFormat(){
```
`getFormat` -> `getFormatFlags`

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16Params.java:
```diff
@@ -0,0 +1,67 @@
+        return format & FORMAT_FLAG;
+    }
+
+    protected int getX32(){
```
These getters should have return types that match whats being tested; i.e., if these are really flags, they should return a boolean.

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16Params.java:
```diff
@@ -0,0 +1,67 @@
+    private int format;
+
+    // To get the least significant 7 bits of format byte
+    private static final int FORMAT_FLAG = 0x7f;
```
`FORMAT_FLAG` -> `FORMAT_FLAG_MASK`

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16Params.java:
```diff
@@ -0,0 +1,67 @@
+    // format is the first byte of the compressed data stream,
+    // which consists of all the bit-flags detailing the type of transformations
+    // and entropy encoders to be combined
+    private int format;
```
`format` -> `formatFlags`

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16Params.java:
```diff
@@ -0,0 +1,67 @@
+    private static final int RLE_FLAG = 0x40;
+    private static final int PACK_FLAG = 0x80;
+
+    public RANSNx16Params(int format) {
```
`format` -> `formatFlags`

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSParams.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface RANSParams {
+
+    public enum ORDER {
```
`public` is redundant in interfaces, and can be removed.

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    public ByteBuffer uncompress(final ByteBuffer inBuffer, int n_out) {
```
We generally try not to use arguments (`n_out`) as mutable local variables. I would make it final, and use a local to derive the actual value used based on 'nosz'.

cmnbroad replied:
Also, if this argument (`n_out`) is the only `uncompress` arg besides the buffer, I think it would be fine unify the `uncompress` method across 4x8 and Nx16 so they have the same signature, and just have the 4x8 implementation ignore this value. The the `uncompress` method can be moved in to the `RANSNew` interface.

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/RANSNx16.java:
```diff
@@ -0,0 +1,152 @@
+        if (nosz == 0) {
+            n_out = Utils.readUint7(inBuffer);
+        }
+        ByteBuffer outBuffer = ByteBuffer.allocate(n_out);
```
It seems like the common case is that `n_out` is zero, in which case is it appropriate/intentional that all of these buffers here and below are allocated as 0 length ?

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +50,173 @@ public String toString() {
+    @DataProvider(name="ransCodecsAndData")
+    public Object[][] getRansCodecsAndData() {
+        // this data provider provides all the testdata for each of the RANS type and order
+        return TestNGUtils.cartesianProduct(getRansTestData(), getRansCodecs());
```
It seems inconsistent that `getRansCodecs` is annotated as a data provider and `getRansTestData` isn't. They should both the the same (??), probably not annotated.

cmnbroad replied:
Also, I do think that, given that the `compress` signature and the data types of the arguments to the `compress` methods are different for 4x8 and Nx16, we'll probably need to separate this dataprovider into two separate dataproviders, one that does the cartesian product of all of the 4x8s test cases (with varying orders, etc) with the test data, and one that does the cartesian product of Nx16s with the test data. Then you'll have to have different (separate) tests for those two data providers, each one of which will have arguments of the correct corresponding type.

cmnbroad replied:
Actually, it more or less looks like thats what you have in the CRAMCorpus tests (separate providers and tests for 4x8 and Nx16).

yash-puligundla replied:
As of the next commit, I am annotating a method as a data provider only if it is used by at least one test as a data provider.

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +50,173 @@ public String toString() {
-    @Test
+    @Test()
     public void testBuffersMeetBoundaryExpectations() {
+        // TODO: Refactor
```
Yeah, once we get the pattern established for these, hopefully we can refactor this to eliminate the redundant code; certainly the ORDER values  can easily be lifted to a data provider. And we'll probably need to separate the 4x8 and Nx16 cases into separate data providers and test methods as suggested for the other tests.

cmnbroad reviewed (changes requested):
Checkpointing what I have so far. I think we should seriously consider separating the encoder and decoder per our slack discussion. In addition to simplifying the code, it will be a win for decoding efficiency since it won't need to allocate all of the encoding machinery it currently does.

Also, I would definitely suggest to focus on getting the test code in shape as the top priority, before staring on the separation of the encoder and decoder.

yash-puligundla commented:
New changes -

- separate encoder, decoder classes
- replaced old rans with rasn4x8
- refactored ranstest.java
- moved uncompress method into the abstract class

yash-puligundla commented:
major changes in the last commit:
- separate packages for rans4x8 and ransnx16
- de-duplicate ransTests code

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
+            final RANSDecode ransDecode,
+            final RANSParams params){
+        final ByteBuffer in = ByteBuffer.wrap(tc.testArray);
+        for (int size = lowerLimit; size < upperLimit; size++) {
```
I know you didn't add this, but do you understand what the point of these limits are ? It seems redundant to repeatedly do the test over different size fragments of the buffer. Can you see any value to that ?

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
-        roundTripForEachOrder(tc.testArray);
+    public Object[][] getRansTestDataTinySmallLarge() {
+        return new Object[][]{
+                { new TestCaseWrapper(randomBytesFromGeometricDistribution(100, 0.1)), 1, 100 }, // Tiny
```
The `TestCaseWrapper` used to be a test **case** wrapper; now its just a test **data** wrapper. Maybe it should be renamed to that, or `TestDataEnvelope`.

cmnbroad replied:
Also, can you add comments to these data providers above the list of values, saying whats in the returned array, like `// test data, lower limit, upper limit` or something similar.

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
+            final Integer upperLimit,
+            final RANSEncode ransEncode,
+            final RANSDecode ransDecode,
+            final RANSParams params){
```
I would suggest to make the signatures of all of the test methods similar wherever possible, i.e., always use the order `encoder, decoder, params, test data, other stuff...`). Your hands are somewhat tied by the results of the cartesian product code, but I think if you can just reverse the order of the params to the call to `cartesianProduct`, the result would read more easily.

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
+        return new Object[][]{
+                { new TestCaseWrapper(randomBytesFromGeometricDistribution(100, 0.1)), 1, 100 }, // Tiny
+                { new TestCaseWrapper(randomBytesFromGeometricDistribution(1000, 0.01)), 4, 1000 }, // Small
+                { new TestCaseWrapper(randomBytesFromGeometricDistribution(100 * 1000 + 3, 0.01)), 100 * 1000 + 3 - 4, 100 * 1000 + 3 } // Large
```
Would be great to document what these magic numbers mean (3, 4 ?) , if anything. It seems like there is some relationship between the args to `randomBytesFromGeometricDistribution` (since they appear there as well). Not sure if you know, but ideally we should figure it out and give them names.

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
+        final ByteBuffer compressed = ransBufferMeetBoundaryExpectations(size,ransEncode, ransDecode,params);
+        Assert.assertTrue(compressed.limit() > 10);
+        Assert.assertEquals(compressed.get(), (byte) params.getOrder().ordinal());
+        Assert.assertEquals(compressed.getInt(), compressed.limit() - 1 - 4 - 4);
```
Same here - can we make constants with symbolic names for whatever the 1 and the 4 represent ?

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
+            final RANS4x8Params params) {
+        final int size = 1001;
+        final ByteBuffer compressed = ransBufferMeetBoundaryExpectations(size,ransEncode, ransDecode,params);
+        Assert.assertTrue(compressed.limit() > 10);
```
`> 10`? If `size` is really `limit`, then should this really be 1001 ?

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -49,107 +57,145 @@ public String toString() {
+            final RANSDecode ransDecode,
+            final RANSParams params) {
+        final ByteBuffer compressed = ransEncode.compress(data, params);
+        final ByteBuffer uncompressed = ransDecode.uncompress(compressed, params);
```
Again, I think you didn't write this, but isn't the while loop below just a long winded way of saying of `Bytebuffer.equals`. Probably have to check  the `ByteBuffer` contract is, but it seems like it would be much less verbose.

cmnbroad commented on src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java:
```diff
@@ -160,8 +206,31 @@ private static void roundTripForOrder(final ByteBuffer data, final RANS.ORDER or
+        final ByteBuffer raw = ByteBuffer.wrap(randomBytesFromGeometricDistribution(size, 0.01));
+        final ByteBuffer compressed = ransEncode.compress(raw, params);
+        final ByteBuffer uncompressed = ransDecode.uncompress(compressed,params);
+        Assert.assertFalse(compressed.hasRemaining());
```
The asserts should be immediately after the code they're checking, i.e., presumably this is testing that the compress call consumed all the data - if so, it should be moved to right after the compress call. No sense in uncompressing if the compress criteria weren't met.

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/Utils.java:
```diff
@@ -34,19 +34,110 @@ static void reverse(final ByteBuffer byteBuffer) {
 
     // Returns the current cumulative frequency (map it to a symbol yourself!)
-    static int RANSDecodeGet(final int r, final int scaleBits) {
+    public static int RANSDecodeGet(final int r, final int scaleBits) {
```
If this returns the cumulative frequency, let's rename it to `getCumulativeFrequency`. (Likewise `r` could use a better name - its the byte for the nth interleave window right ? Maybe `ransByte`.)

cmnbroad commented on src/main/java/htsjdk/samtools/cram/compression/rans/Constants.java:
```diff
@@ -1,7 +1,9 @@
+    public static final int TF_SHIFT = 12;
+    public static final int TOTFREQ = (1 << TF_SHIFT); // 4096
+    public static final int RANS_BYTE_L_4x8 = 1 << 23;
+    public static final int RANS_BYTE_L_Nx16 = 1 << 15;
```
`RANS_BYTE_L...` represents the lower bound for renormalization ? Could use a better name.

cmnbroad reviewed (changes requested):
Another round of comments for discussion. WIP.

yash-puligundla commented:
@cmnbroad Thanks for the feedback. I have addressed all your comments so far. The PR is ready for the next round.
